src/Domain/TagsCharacteresticSpeeds.hpp

```diff
@@ -0,0 +1,54 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <boost/optional.hpp>
+#include <cstddef>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/EagerMath/Magnitude.hpp"  // For Tags::Normalized
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Domain/Tags.hpp"
+#include "Domain/TagsTimeDependent.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace domain {
+namespace Tags {
+/// Compute the characteristic speeds on the moving mesh given the
+/// characteristic speeds if the mesh were stationary.
+///
+/// \note Assumes that `typename CharSpeedsComputeTag::return_type` is a
+/// `std::array<DataVector, NumCharSpeeds>`
+template <typename CharSpeedsComputeTag, size_t Dim>
+struct CharSpeedCompute : CharSpeedsComputeTag {
+  using base = CharSpeedsComputeTag;
+  using return_type = typename CharSpeedsComputeTag::return_type;
+
+  template <typename... Ts, typename T, size_t NumerOfCharSpeeds>
+  static void function(
+      const gsl::not_null<std::array<T, NumerOfCharSpeeds>*> result,
+      const boost::optional<tnsr::I<DataVector, Dim, Frame::Inertial>>&
+          grid_velocity,
+      const tnsr::i<DataVector, Dim, Frame::Inertial>& unit_normal_covector,
```

Comment:
For my understanding: this normal is presumably also one of the arguments `ts` passed in to `CharSpeedsComputeTag::function(result, ts...);` to compute the "original" char speeds, right? You just need to pass it again here for the velocity adjustment?